File: apps/frontend/src/types/habit.ts
export interface Habit {
  id: string;
  title: string;
  frequency: "daily" | "weekly" | "monthly";
  target_count: number;
  color: string | null;
  streak: number;
  created_at: string;
}

export interface HabitLog {
  id: string;
  habit_id: string;
  completed_at: string;
  created_at: string;
}

File: apps/backend/app/models/habit.py
from datetime import datetime

from sqlalchemy import DateTime, String, Integer, ForeignKey, func, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Habit(Base):
    __tablename__ = "habits"

    id: Mapped[str] = mapped_column(Uuid(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid())
    user_id: Mapped[str] = mapped_column(Uuid(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    frequency: Mapped[str] = mapped_column(String(20), nullable=False, default="daily")
    target_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    color: Mapped[str | None] = mapped_column(String(7), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    logs = relationship("HabitLog", back_populates="habit", cascade="all, delete-orphan")
